src/server/DeviceServer.js

```diff
@@ -452,15 +458,17 @@ class DeviceServer {
         device.setOtaChunkSize(Number.parseInt(nullthrows(eventData.data), 10));
       }
 
-      if (
-        eventName.startsWith(SYSTEM_EVENT_NAMES.SAFE_MODE)
-      ) {
+      if (eventName.startsWith(SYSTEM_EVENT_NAMES.SAFE_MODE)) {
         this.publishSpecialEvent(
           SYSTEM_EVENT_NAMES.SAFE_MODE,
           eventData.data,
           deviceID,
           ownerID,
         );
+
+        if (this._enableSystemFirmwareAutoupdates) {
+          await this._updateDeviceSystemFirmware(device);
```

Comment:
Don't you already have `systemInfo` and `deviceID` before this? You could just pass those in instead of querying again.